File: 496-next-greater-element-i/496-next-greater-element-i.cpp
class Solution {
public:
    
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) 
    {
        stack <int> st;
        map <int, int> mp;
        for(int i : nums2)
        {
            while(!st.empty() && st.top() < i)
            {
                mp[st.top()] = i;
                st.pop();
            }
            st.push(i);
        }
            vector <int> ans;
            for(int j = 0; j < nums1.size(); j++)
            {
                if(mp.find(nums1[j])!=mp.end())
                    ans.push_back(mp[nums1[j]]);
                else
                    ans.push_back(-1);
            }
        return ans;
    }
};

File: 496-next-greater-element-i/496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        
        stack = []
        dic = {}
        for i, num in enumerate(nums2):
            while stack and stack[-1] < num:
                dic[stack[-1]] = num
                stack.pop()
            stack.append(num)
        ans = []
        for num in nums1:
            if num not in dic:
                ans.append(-1)
            else:
                ans.append(dic[num])
        return ans
        